098-using-image-volumes-to-improve-extensibility-of-Strimzi-operands.md

```diff
@@ -0,0 +1,335 @@
+```
+
+Once this container image is pushed into the registry, it can be used as a source for the image volume.
+The container image contains only the plugin (and its dependencies).
+It does not contain any operating system artifacts or Strimzi JARs.
+Its rebuild is therefore needed only when a new version of the plugin is released.
+
+It is users responsibility to ensure that the files in the container image will not cause any class path conflicts and break the operand.
+(This is the same already today with the pre-existing mechanisms for adding plugins.)
+
+User is also responsible for securing the container images with the plugins and for securing the container registry.
+When pulling images and using them as volumes, Kubernetes will apply the same checks as with the regular container images used by the operand.
+Using the container images with plugins from untrusted sources might constitute a risk.
+
+### Kafka Connect plugins (connectors, transformations, etc.)
+
+Kafka Connect plugins continue to be handled in their own separate way even with image volumes through a new section in the `KafkaConnect` custom resource.
+This is because they are the most commonly used type of Apache Kafka plugins and deserve the _improved_ user experience.
+But it has its own technical reasons as well:
+* The connector plugins are stored in a separately defined path defined by the `plugin.path` option in Apache Kafka.
+  They are normally not included in the regular class path as most other plugins.
+* Having a separate API allows us to mix the existing approaches (Kafka Connect Build or custom container image) with the connectors mounted from image volumes.
+  This might be useful for example when some plugins are not (yet) available as container images.
+
+The plugins will be configured in the new section in `.spec.plugins`.
+The structure of this section is identical to the existing `.spec.build.plugins` used to define the plugins that should be added through the Kafka Connect Build feature.
+But the artifact types will be different.
+
+The `.spec.plugins` section will contain a list of plugins.
+Each plugin will consist of one or more artifacts to allow compose the plugin from different libraries if needed.
+As of this proposal, the only artifact type will be `image`.
+It will let users specify the reference to the container image with the plugin that will be mounted and used.
+The image can be specified using a tag or using a digest.
+The advantage of the digest is that it exactly specifies the container image that will be used.
+The advantage of the tag is that you can always use the latest version of the plugin without updating the custom resource.
+But it will not be deterministic which exact version is used.
+
+The following example shows the new YAML of the `KafkaConnect` resource:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaConnect
+metadata:
+  # ...
+spec:
+  # ...
+  plugins:
+    - name: echo-sink-connector
```

Comment:
Once [KIP-891](https://cwiki.apache.org/confluence/display/KAFKA/KIP-891:+Running+multiple+versions+of+Connector+plugins) is accepted it will be possible to have multiple versions of a connector involved. I wonder if we should incorporate this into the proposal at all. Maybe it is enough that people can just use different names for the different versions if they want to have multiple versions of the same plugin, but thought it was worth considering.

Reply:
We are pretty indifferent to the version or to the actual content. So if we add a version field, all we would do with it is incorporate it into the path. But Connect will anyway use the version it reads from the plugin. So I think it would not be very useful and when using multiple versions of the same plugin, users can incorporate the version into the name I guess?

Reply:
I think the user incorporating the version into the name is enough